### ISSUE TITLE
Q5 Can't see difference between stepAIC() and step()

### ISSUE THREAD

**Issue by frie-knuth** [dmcglinn/quant_methods#4]
For some reason the forward and backward outputs for `stepAIC()` and `step()` are identical. Not sure what I am missing...

Here is what I ran in Q2 to get me up to Q5

```{r}
## load data
trees = read.csv('./treedata.csv')
## count species in each plot 
richness = with(trees, tapply(species, plotID, FUN = function(x) length(unique(x))))
## convert richness to data frame
rich_dat = data.frame(plotID = names(richness), plot_rich = richness)
## merge rich_dat with trees
trees = merge(trees,rich_dat, by="plotID", all=TRUE)
```

Cool. Now in Q5 I ran the following:

```{r}
## create glm model following poisson distribution
mod_rich = glm(plot_rich ~ utme + utmn + elev + tci +streamdist + disturb + beers, family = poisson, data = trees)

##run forward and backward stepAIC() and step() functions.
stepAIC(mod_rich, direction = "backward")
stepAIC(mod_rich, direction = "forward")
step(mod_rich, direction = "backward")
step(mod_rich, direction = "forward")
```

Again, the outputs for `stepAIC(mod_rich, direction = "backward")` and `step(mod_rich, direction = "backward")` as well as `stepAIC(mod_rich, direction = "forward")` and `step(mod_rich, direction = "forward")` are the same, but I have a hunch they shouldn't be.

**Comment by dmcglinn** [comment]
I'll have to look into this. My memory may be lagging here but I though that `step()` used p-values to chose the best model and `stepAIC()` used the AIC metric to chose the best model. It's possible they are both using AIC in which case you wouldn't expect differences. Also with respect to forward and backward model selection these not not guaranteed to give the same results, but I suppose with the small number of variables we are considering in the trees dataset that this is very likely.

**Comment by dmcglinn** [comment]
Oh and your line above
```
trees = merge(trees, rich_dat, by="plotID", all=TRUE)
```
has a flaw in it. Run `nrow(rich_dat)` and `nrow(trees)`. You want your end result to be the same number of rows as `rich_dat`. Hmm now how would you fix that? Maybe take a look back at the arguments all, all.x, and all.y.

**Comment by knuthfa** [comment]
I tried all combinations of all, all.x and all.y = TRUE and FALSE just to see what happens and I always end up with 8971 rows, instead of 935. But in a sense, if I just wanted the environmental data for each plot associated (merged) with the species count for each plot, resulting in 935 rows, I could just merge the product of Q2 with the product of Q4 before going on to Q5. However, when I do this, multiple environmental variable cells contain more than one enter i.e. a list, and are thus invalid inputs into the glm. What am I missing here?

**Comment by knuthfa** [comment]
...contain more than one *entry*...

**Comment by dmcglinn** [comment]
This is command works
```
rich_dat = data.frame(plotID = names(richness), plot_rich = richness)
rich_env = merge(rich_dat, trees, all.x=T, all.y=F)
```

Note above I did not specify the "by" argument. For some reason that was messing things up but it really shouldn't have been.

**Comment by knuthfa** [comment]
I still end up with 8971 rows... But even if I end up with 935, will it simply have concatenated environmental variable info in the other columns? How will the glm() function in Q5 handle concatenated lists as a value for a given x?

**Comment by dmcglinn** [comment]
You should recheck your code and consider restarting R. When you use `merge(x,y)` it will add any column from x and y that does not have the same name into an output data.frame. It will match rows based on any column names they share in common. If you were to try to model the 8971 rowed data.frame then the GLM would assume that you had many more degrees of freedom (i.e., samples) then you really have (i.e., 935) so all p-values will be more liberal than they should be.

**Comment by knuthfa** [comment]
restarted R - rechecked code. No change. Not sure what is up. I understand how a merge works from using ArcGIS and also roughly understand why modeling 8971 with redundant entries for plot richness is a problem.

I may have to show you in person what the issue is I am seeing, even if the merge works. I got the merge to work using a workaround by merging the output from Q4 and Q2, which gives me the desired datable with 935 entries. This however won't work as an input for the `glm()` The problem is that a single plot may have multiple streamdist, tci, disturb etc. values associated with it. These are expressed as a concatenated list within a single x,y cell within the merged 935 cell data.frame.

Sorry if this isn't clear. Again, I might just have to show you in person. Am pretty much ready to submit, which might be the best solution at this point for you to see where I am going wrong.

Thanks for all your help!

**Comment by dmcglinn** [comment]
sounds good. go ahead and submit it then.

**Comment by dmcglinn** [comment]
This question was finally resolved in #6 I was mistaken in how the function `merge` behaves.
